drivers/pwm/pwm_nrfx.c

```diff
@@ -31,36 +31,59 @@ struct pwm_nrfx_config {
 
 struct pwm_nrfx_data {
 	uint32_t period_cycles;
-	uint16_t current[NRF_PWM_CHANNEL_COUNT];
-	uint16_t countertop;
+	uint16_t seq_values[NRF_PWM_CHANNEL_COUNT];
 	uint8_t  prescaler;
 	uint8_t  initially_inverted;
+	bool     stop_requested;
 };
 
 
-static int pwm_period_check_and_set(const struct pwm_nrfx_config *config,
-				    struct pwm_nrfx_data *data,
-				    uint32_t channel,
-				    uint32_t period_cycles)
+static bool channel_needs_pwm(uint32_t channel,
+			      const struct pwm_nrfx_data *data)
+{
+	uint16_t compare_value =
+		data->seq_values[channel] & PWM_NRFX_CH_COMPARE_MASK;
```

Comment:
is `seq_values` being used only to determine if pwm is used? If yes then it would be simpler to store bit flags. Check would be faster, less ram, etc.

Reply:
No, the values determine pulse widths for particular channels. See this part:
https://github.com/zephyrproject-rtos/zephyr/blob/89928afbe8ef0eb4a6a4c45146001f112c1ec800/drivers/pwm/pwm_nrfx.c#L168-L172

For channels that are set to constant inactive or active level we have here 0 or `PWM_NRFX_CH_COMPARE_MASK`, respectively, what means that such channel does not need to be driven by the PWM peripheral, and that's what the `channel_needs_pwm` function checks.

Reply:
i didn't notice that it's used by the nrfx driver by assigning `.seq.values.p_raw = pwm_nrfx_##idx##_data.seq_values,`
but still, maybe it would be worth to create `uint8_t` mask in the data structure (it would fit without increasing the size) and track use of pwm there. Check would be simplified to just mask against 0.

Reply:
I'm not convinced it would actually be an improvement. This information would need to be maintained in two places then. I still need to provide proper values in the `seq_values` array for those constant channels. Although they do not need the PWM peripheral, they will still be driven by it when the peripheral is required for any other channel.

Reply:
i won't insist. Though, imo it would be. There is one if/else which would set/clear flags: https://github.com/zephyrproject-rtos/zephyr/blob/89928afbe8ef0eb4a6a4c45146001f112c1ec800/drivers/pwm/pwm_nrfx.c#L152 and then `any_other_channel_needs_pwm` become instant.

Reply:
Okay, now I see it (a bit late, though :wink:). I'll improve this in a follow-up PR.

Reply:
Addressed in #48353.